Add ability to compile generated code during testing

arporter opened the pull request:
This PR is for work done under #32.

arporter commented:
Hi @rupertford, not sure whether we should merge this on before or after our freeze for pypi.

arporter commented:
@rupertford I've brought the associated branch up-to-date with master so that it now has the new directory structure (this then lets me use it in other branches without too much pain).

arporter commented:
Realised that we can add command-line options to py.test using `addopts`. I have therefore set-up some fixtures in `conftest.py`. A test that involves compilation should now have the `@utils.compile` decorator. Such tests are then only performed if the `--compile` flag is passed to py.test. The default compiler is `gfortran` but this, and the compile flags, can be changed using the `--f90=my_compiler` and `--f90flags="some flags"` options.

As part of this change I have removed the look-up of the F90 and F90FLAGS environment variables.

rupertford commented on src/psyclone/tests/utils.py:
```diff
@@ -28,3 +50,141 @@ def count_lines(root, string_name):
+                                 "test_files", "dynamo0p3")
+        from dynamo0p3_build import INFRASTRUCTURE_MODULES as module_files
+
+    kernel_modules = []
```
I think kernel_modules can be found simply by psy_gen.invokes.names. This would remove the need for the psy_module change to dynamo0p3.py and psyGen.py and the walk function above

arporter replied:
Well spotted. It's not quite as simple as you suggest - I have to dig down to the kernel objects and then get their associated `module_name` - but it's nicer.

rupertford commented on src/psyclone/tests/utils.py:
```diff
@@ -28,3 +50,141 @@ def count_lines(root, string_name):
     return count
+
+
+def walk(parent, my_type):
```
I think this can be removed if the changes proposed below are possible (see next comment).

arporter replied:
Removed.

rupertford commented on src/psyclone/tests/utils.py:
```diff
@@ -28,3 +50,141 @@ def count_lines(root, string_name):
+        return True
+
+
+def code_compiles(api, psy_ast, tmpdir, f90, f90flags):
```
If we can integrate the build tests with existing correctness tests then it might be better to pass in the code as a string here rather than the ast (psy_ast), as it will already have been generated within the test. If we are going for different correctness and compile tests then there is no need to change.

arporter replied:
Now that we look for USE'd modules using the `psy_ast` we cannot replace it with a string. In fact this means that my optimisation to store the generated AST is now valid again but, as it is an optimisation, I'll leave it as it is for now.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -922,27 +922,36 @@ def gen(self):
+        # return the AST for the generated code
+        return self.psy_module.root
+
+    @property
```
Agreed - I've removed it.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -214,6 +214,7 @@ def __init__(self, invoke_info):
 
         self._name = invoke_info.name
         self._invokes = None
+        self._psy_module = None
```
Removed.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -231,6 +232,11 @@ def gen(self):
         raise NotImplementedError("Error: PSy.gen() must be implemented "
                                   "by subclass")
 
+    @property
```
Removed.

arporter commented on src/psyclone/tests/dynamo0p3_build.py:
```diff
@@ -0,0 +1,23 @@
+''' Module containing configuration required to build code generated
```
Oh yes - added.

arporter commented:
I've added `utils.COMPILE_TEST` which is set according to whether or not the `--compile`` flag is passed to py.test. This can then be used within existing tests to decide whether or not to test the compilation of the generated code. I've modified the two existing examples of testing compilation to use this rather than having duplicate tests.
I feel the need to document this new feature but this is very much something for a developer and not a user. I could begin a new Developing chapter in the existing documentation?

As suggested, I've removed all extraneous comments and code from the infrastructure stubs (including one file that I'd duplicated!). All copyright/license notices have been modified so that the new MO text comes first followed by ours. I think I've addressed all other comments.

Back to @rupertford for 2nd review...
